feat: add a KubernetesContextState component

feloy opened the pull request:
Signed-off-by: Philippe Martin <phmartin@redhat.com>

### What does this PR do?

This PR adds a KubernetesContextState to the backend. 

This component watches pods and deployments (using informers) on all the contexts defined in the user's kubeconfig file.

Every time a change happens in any of the context (number of pods/deployment, context is reachable or not), the new context is sent with a message `kubernetes-contexts-state-update`

Part of #4562 

The PR #5917, based on this one, can be used to test with the information displayed on the frontend

benoitf commented on packages/main/src/plugin/testutils/fake-informer.ts:
```diff
@@ -0,0 +1,64 @@
+
+import type { ErrorCallback, KubernetesObject, ObjectCallback } from '@kubernetes/client-node';
+
+export class FakeInformer {
```
hello, is that this class will be used in other places than kubernetes-context-state ?

feloy replied:
For this feature no, but it is very generic and could be useful for any other code using informers

benoitf replied:
I would keep the objet in the kubernetes-context-state.spec.ts and only if we need it elsewhere we would move it rather than creating a `testutils` directory with an hypothetical reuse of this object.

Also

deboer-tim commented on packages/main/src/plugin/kubernetes-context-state.ts:
```diff
@@ -0,0 +1,223 @@
+    };
+  }
+
+  private createPodInformer(kc: KubeConfig, ns: string, context: KubeContext): Informer<V1Pod> & ObjectCache<V1Pod> {
```
There is a lot of duplication between createPodInformer and createDeploymentInformer. Could you refactor out a common function sort of like kubernetes-client.ts makeKubernetesInformer?

dgolovin replied:
There are many smaller duplications like:

```
      const previous = this.contextsState.get(context.name);
      if (previous) {
        /*
           do something with previous context state
        */
      }
      this.dispatchContextsState();
```

deboer-tim reviewed (commented):
Sorry for this long review. I'm not sure why it wasn't fully dawning on me before (sorry Philip, I think you've been telling me), but after this PR we will have:

1. pod data pulled using list + watcher.
2. deployments/services/ingress/routes list + informers for current context, with state managed in the store via add/remove events.
3. pod/deployment counts pulled for all contexts by these informers, with future store.

To me #1 is unchanged technical debt; we didn't want to risk a failed release while trying informers/Kubernetes UI, but we've been using them long enough now to know they're better in all respects.

I like that #2 gives us add/remove events in the renderer, but we may never need it and could replicate. Our long-term design mockups had the ability to see resources in multiple contexts at once, and #3 is the only multi-context pattern.

So accepting this PR will create some overlapping function for the current context and technical debt, but I also think it gives us a better way to get to multiple contexts without any disruption, e.g. a path something like:

- Merge this PR (which would unblock Settings > Kubernetes UI improvements, and ability to put current connection status on Deployments/Services/etc pages or statusbar)
- Update ContextsState to keep arrays for pods & deployments instead of counts.
- Add services/ingress/routes to ContextsState.
- Change kube resource stores to use ContextsState.
- Delete 'old' informers.
- Update pods to use informer store, delete old pod watcher.

Thoughts? My only real concern with the existing code was to reduce duplicate code in createXInformer functions.

feloy commented on packages/main/src/plugin/kubernetes-context-state.ts:
```diff
@@ -0,0 +1,230 @@
+    return this.createInformer<V1Deployment>(kc, context, path, listFn, 'deploymentsCount', { checkReachable: false });
+  }
+
+  private createInformer<T extends KubernetesObject>(
```
This method may be moved to kubernetes-client.ts sooner or later

feloy commented:
> Sorry for this long review. I'm not sure why it wasn't fully dawning on me before (sorry Philip, I think you've been telling me), but after this PR we will have:
> 
> 1. pod data pulled using list + watcher.
> 2. deployments/services/ingress/routes list + informers for current context, with state managed in the store via add/remove events.
> 3. pod/deployment counts pulled for all contexts by these informers, with future store.
> 
> To me #1 is unchanged technical debt; we didn't want to risk a failed release while trying informers/Kubernetes UI, but we've been using them long enough now to know they're better in all respects.
> 
> I like that #2 gives us add/remove events in the renderer, but we may never need it and could replicate. Our long-term design mockups had the ability to see resources in multiple contexts at once, and #3 is the only multi-context pattern.
> 
> So accepting this PR will create some overlapping function for the current context and technical debt, but I also think it gives us a better way to get to multiple contexts without any disruption, e.g. a path something like:
> 
> * Merge this PR (which would unblock Settings > Kubernetes UI improvements, and ability to put current connection status on Deployments/Services/etc pages or statusbar)
> * Update ContextsState to keep arrays for pods & deployments instead of counts.
> * Add services/ingress/routes to ContextsState.
> * Change kube resource stores to use ContextsState.
> * Delete 'old' informers.
> * Update pods to use informer store, delete old pod watcher.
> 
> Thoughts? My only real concern with the existing code was to reduce duplicate code in createXInformer functions.

I have refactored the createXInformer with a single generic createInformer. Handlers are passed to this method, to make it completely agnostic of its context, to make it (hopefully) reusable in other parts of the code.

deboer-tim commented on packages/main/src/plugin/kubernetes-context-state.ts:
```diff
@@ -0,0 +1,227 @@
+    });
+    informer.on('error', (err: unknown) => {
+      if (err !== undefined) {
+        console.error(`pod informer error for context ${context.name}: `, String(err));
```
'pod' needs to be replaced here, maybe use path in the message instead

deboer-tim commented on packages/main/src/plugin/kubernetes-context-state.ts:
```diff
@@ -0,0 +1,227 @@
+    if (options.onReachable) {
+      informer.on('connect', (err: unknown) => {
+        if (err !== undefined) {
+          console.error(`pod informer connect error for context ${context.name}: `, String(err));
```
Ditto here.

deboer-tim reviewed (approved):
Thanks for the changes, just a couple minor issues noted.

feloy commented:
> Thanks for the changes, just a couple minor issues noted.

Fixed, thanks

deboer-tim reviewed (approved):
Adding back my approval, changes were addressed.

feloy commented:
> In my first PR (#5329), I had added a Preference to enable/disable running informers on all clusters of the kubeconfig files, and disabled by default, because I was afraid that user with many old/unexisting clusters on the kubeconfig would retry connecting to all of them infinitely.
> 
> I removed this preference in this PR, because I understand that we would like to reuse the information, at least fot the current context.
> 
> But I'm still wondering if we could have this preference for the non-current context, to avoid trying to connect to these non-current contexts.
> 
> @deboer-tim @benoitf WDYT?

As discussed in daily meeting, I'll add an exponential backoff with jitter

benoitf commented on packages/main/src/plugin/kubernetes-context-state.ts:
```diff
@@ -0,0 +1,227 @@
+// of instances of different resources
+export interface ContextState {
+  reachable: boolean;
+  podsCount: number;
+  deploymentsCount: number;
```
it's just questions as it's an internal object we can changes later, when we want

should it be podsCount or `{pods: {count : number}}` if we add other stuff related to pods later

about reachable, I think we should add the last time it was reachable as well

deboer-tim replied:
IMHO it should be changed to `pods: PodInfo[]` in a future PR, ditto for deployments. Easy enough to get the count from there and can replace current-context informer.

Good call on reachable. We could track timestamp on every interaction so that when disconnected we know the last time it was connected, and when connected we know the last time we had an update or when it connected. IMHO this could wait though.

benoitf commented on packages/main/src/plugin/kubernetes-context-state.ts:
```diff
@@ -0,0 +1,227 @@
+    let removed = false;
+    for (const [name, state] of this.contextsInternalState) {
+      if (!this.kubeConfig.contexts.find(c => c.name === name)) {
+        await state.podInformer?.stop();
+        await state.deploymentInformer?.stop();
```
should we have only a stop method that does that for us, instead of the need to call sub parts

in addition, It's not also mapping overall arch that is using disposable pattern

so when I register something I have a disposable object that will do stop, delete, etc
cleaning is just calling the dispose() method on the disposable object

so I would have only called dispose() on the state and it would have self stopped/removed the resource

feloy replied:
it should be fixed with latest commit (refactor: use dispose pattern for state)

benoitf commented on packages/main/src/plugin/kubernetes-context-state.ts:
```diff
@@ -0,0 +1,227 @@
+      clearTimeout(this.timeoutId);
+    }
+
+    this.timeoutId = setTimeout(() => {
+      this.apiSender.send(`kubernetes-contexts-state-update`, this.contextsState);
+    }, 100);
+  }
```
the timeout looks like too low (or high depending on how we look)
I guess we could be like 1second

benoitf commented on packages/main/src/plugin/kubernetes-context-state.ts:
```diff
@@ -0,0 +1,227 @@
+      }
+      options.onReachable?.(err === undefined);
+      // Restart informer after 5sec
+      setTimeout(() => {
+        this.restartInformer<T>(informer, context, options);
+      }, 5000);
```
where is kept that setTimeout reference ?

if there is like several errors, would it add two times restartInformer ?

feloy replied:
I have added a commit with timer related changes

benoitf commented on packages/main/src/plugin/kubernetes-context-state.ts:
```diff
@@ -0,0 +1,227 @@
+  ) {
+    informer.start().catch((err: unknown) => {
+      if (err !== undefined) {
+        console.log('informer start error: ', String(err));
```
```suggestion
        console.error('informer start error: ', String(err));
```
(or warn but probably not console.log)

benoitf commented on packages/main/src/plugin/kubernetes-context-state.ts:
```diff
@@ -0,0 +1,227 @@
+        console.log('informer start error: ', String(err));
+      }
+      options.onReachable?.(err === undefined);
+      // Restart informer after 5sec
+      setTimeout(() => {
+        this.restartInformer<T>(informer, context, options);
+      }, 5000);
```
same question than before on the reference to the timeout reference

benoitf commented on packages/main/src/plugin/testutils/fake-informer.ts:
```diff
@@ -0,0 +1,64 @@
+
+  constructor(
+    private resourcesCount: number,
+    private connectResponse: Error | undefined,
+  ) {
+    this.onCb = new Map<string, ObjectCallback<KubernetesObject>>();
+    this.offCb = new Map<string, ObjectCallback<KubernetesObject>>();
+    this.onErrorCb = new Map<string, ErrorCallback>();
+  }
+  async start(): Promise<void> {
+    this.onErrorCb.get('connect')?.(this.connectResponse);
+    if (this.connectResponse === undefined) {
+      for (let i = 0; i < this.resourcesCount; i++) {
+        this.onCb.get('add')?.({});
+      }
+    }
+  }
+  stop() {}
+  on(
+    verb: 'change' | 'add' | 'update' | 'delete' | 'error' | 'connect',
+    cb: ErrorCallback | ObjectCallback<KubernetesObject>,
+  ) {
+    switch (verb) {
+      case 'error':
+      case 'connect':
+        this.onErrorCb.set(verb, cb as ErrorCallback);
+        break;
+      default:
+        this.onCb.set(verb, cb as ObjectCallback<KubernetesObject>);
+    }
+  }
+  off(
+    verb: 'change' | 'add' | 'update' | 'delete' | 'error' | 'connect',
+    cb: ErrorCallback | ObjectCallback<KubernetesObject>,
+  ) {
+    this.offCb.set(verb, cb);
+  }
+  get() {}
+  list() {}
+}
```
I think I would have picked up mocking style here rather than declaring empty method

instead of `get()` --> `get: vi.fn()`

for off, usage of mockImplementation etc

benoitf commented on packages/main/src/plugin/kubernetes-context-state.ts:
```diff
@@ -0,0 +1,295 @@
+
+  private getJitter(): number {
+    // This is not security-related code
+    return Math.floor(this.jitter * Math.random()); // NOSONAR
```
for disabling sonar rules you can do that from the UI of sonar source

but maybe you could just use
```ts
const uintArray = new Uint8Array(1)
const randomInt = crypto.getRandomValues(uintArray)[0]
const randomFloat = randomInt / Math.pow(2, 8)
```

feloy replied:
> for disabling sonar rules you can do that from the UI of sonar source
> 

OK, I'll check agin the UI. Thanks

> but maybe you could just use
> 
> ```ts
> const uintArray = new Uint8Array(1)
> const randomInt = crypto.getRandomValues(uintArray)[0]
> const randomFloat = randomInt / Math.pow(2, 8)
> ```

As it is really not crypto code, I would prefer to keep it simple

lstocchi reviewed (commented):
Not sure if i'm doing something wrong. I merged this PR to https://github.com/containers/podman-desktop/pull/5917. 
I start desktop and log to the sandbox cluster.
Now if i go to the context page i see unreachable to all contexts, but i was expecting to be connected to the sandbox one. To verify i'm connected i also go to the pods page and i see the pod i have there. Then i switch to minikube which is turned off and i don't see any pod as expected.
Am i doing something wrong?
![pod_context](https://github.com/containers/podman-desktop/assets/49404737/fcae863c-510c-44d1-8d6b-eaaec12e6345)

lstocchi reviewed (approved):
Tested by using the new rebased #5917 . Works fine now. The pods/deploy count gets updated correctly as well. LGTM

![image](https://github.com/containers/podman-desktop/assets/49404737/43e0d07d-8f07-4986-a27e-6b55556fb692)
